### PR TITLE
add support for Byte<->raw IO

### PR DESCRIPTION
**Pull request by mdsumner**
I'd like to be able to read and write Byte with R's raw vector type. This is unproblematic at the Rcpp level afaics. 

There are implications for the documentation and plotting (there's a failing plot test). 

I consider this a draft until those implications are discussed and dealt with. Thanks!   

The motivation is purely for the size of Byte vs Int32, for workflows that just move pixel data around. 

--------------------------------------------
Longer term motivation: 

It's not particularly  efficient to get raw vectors and then convert them to colours in R, but in time perhaps we can add helpers for that, i.e. I would also like to add a special r-only  type on input that bundles (1, or 3, or 4) raw vectors into Int32 with R's nativeRaster encoding i.e. [vapour::gdal_raster_nara](https://github.com/hypertidy/vapour/blob/5e67be7626051b3f444b7fec49d829cf3ab6444c/inst/include/gdalraster/gdalraster.h#L71), or alternatively to character string colour. nativeRaster is supremely fast for visualization, and we save some memory copying if done before returning to R.  Note also that R now has smaller format allowed for hex string: https://github.com/wch/r-source/commit/0cad8c3fa75f402db464cbfde93284106a12a69f

### CODE REVIEW

**Comment by mdsumner** [comment]
I guess we can just as.integer in plot_raster when `legend = TRUE`, but will be worth doing that upfront so we get the range and not run quantile/min/max etc multiple times. 

Also, I'm now a bit worried about plotting when raw *as greyscale/rgb/rgba* vs when there is a palette, but I guess that's an easy case to pivot on - it's like the RGB case where the legend is ignored. I'll have to explore more!

**Comment by mdsumner** [comment]
Here's an illustration of the size savings with export to Byte (the point of this PR), and the speed we can get by plotting with nativeRaster (maybe a future consolidation). 

https://gist.github.com/mdsumner/1594a0dc5be2de525cff92eb66524983

**Comment by ctoney** [comment]
Nice, thanks for digging into this. Seems very worthwhile and we should definitely implement.

The only problem I see with the current draft is that Byte raster would be read as raw, whether the user wants that or not. Users will want to read Byte data as integer in most cases, so we should keep that as default. Otherwise it would be a breaking change for existing code that expects integer. 

The usual way we might do an optional `readByteAsRaw` won't work in this case, but there's a decent alternative.

With `RCPP_EXPOSED_CLASS`, we cannot provide default values for arguments to class methods, confirmed here: https://stackoverflow.com/questions/68514809/exposing-class-method-with-default-parameter-using-rcpp. Note that we can supply defaults for arguments to C++ free functions exposed via Rccp modules. That is currently done with several stand-alone functions in gdalraster. It's just a limitation with class methods via `RCPP_EXPOSED_CLASS`.

I see this limitation as a trade-off with the current implementation of class `GDALRaster`. I still like the trade-off, but the reasons are a possible separate discussion at some point (e.g., since we're working toward a similar implementation for `GDALVector`). Would be happy to discuss that more, and certainly open to other views on it.

But for now, we can't just add an argument like `readByteAsRaw` in `GDALRaster::read()`, because we can't supply a default, so it would be a breaking change. All arguments to class methods require a value.

A possible solution in this case is to expose a public field on the class with read/write access from R. Rcpp modules has a `.field` method that does this, documented in the vignette on p. 5 (https://cran.r-project.org/web/packages/Rcpp/vignettes/Rcpp-modules.pdf). It's simple to implement, just a public variable in the class and add the corresponding `.field` in `RCPP_MODULE(mod_GDALRaster)`, which starts at line 1434 in `gdalraster.cpp`.  From R it would look something like

```r
ds <- new(GDALRaster, f)
ds$readByteAsRaw <- TRUE  # default is FALSE
r <- ds$read(...)
# or, r <- read_ds(ds, ...)
```

`read_ds()` might also have an optional argument that would trigger the setting (and then unset before returning), so it could be accomplished either way.

I'm thinking this would only have effect when the raster data type is Byte, i.e., probably not implement `as.raw()` coercion for other raster data types and just ignore the setting if not `GDT_Byte`?

>Coercion to raw treats the input values as representing small (decimal) integers, so the input is first coerced to integer, and then values which are outside the range ‘[0 ... 255]’ or are ‘NA’  are set to ‘0’ (the ‘nul’ byte).

Fields could be useful for other things too. Let me know what you think. I can add a few comments on the rest of the PR once we decide on that.

**Comment by ctoney** [comment]
BTW, the file `inst/extdata/storml_tcc.tif` is Byte, in case you need one for tests. It has a color table, used in `test-display.R`.

**Comment by mdsumner** [comment]
Ah yes, I see - all good. Thanks for the explanation. Happy with that, I think having both pathways on the dataset and via `read_ds()` is good. I actually think it's better overall, having it opt-in like this, because you're right there's lots of situations where red,green,blue as Byte or Int16 or Float will being used and you wouldn't want this detail hampering that.

**Comment by ctoney** [comment]
I agree, we should just `as.integer` for now in `plot_raster`. I think we have to do `as.integer` regardless of whether `legend = TRUE` though. The data go through `.normalize()` which uses `min()` and `max()`. I'm pretty sure all that's needed is add

```r
if (typeof(data_in) == "raw")
    data_in <- as.integer(data_in)
```

at line 369 in `display.R`, right before `a <- array(...)`. The test passed when I did that. Of course, the test would also pass once the other changes discussed above are made, by opting not to read as raw. But this should handle the case of passing raw vector, or passing a dataset object with the new `readByteAsRaw` setting. The plot of `inst/extdata/storml_tcc.tif` is correct when read as raw after making that change.

**Comment by mdsumner** [comment]
Ok cool, let me explore and make that change for `readByteAsRaw`. I appreciate now the real limitations of the Rcpp module approach (but also its convenience!).  I think it's good to stick with that kind of change for now.

**Comment by mdsumner** [comment]
oh nice!  that is some funny magic, you've taught me a lot today, I had tried to get into modules a few times ...

**Comment by mdsumner** [comment]
todo/wip/questions for me to take this out of draft

- document the new field, check documentation for consistency in read_ds
- tests
- consider warning at read time (maybe with a limit on warnings) when it's set by the user but GDT_type is not Byte
- do we have to initialize the field explicitly to false, or can we assume that's safe as is?
-

**Comment by ctoney** [comment]
I think it would safer to initialize `readByteAsRaw` to `false` even though it appears to get default initialized that way. It could be in the initializer list, but the way the constructors are set up currently, we'll have to do it in two places. Can add `readByteAsRaw(false)` to the existing lists for the default constructor at line 78, and also the final constructor that has three arguments at line 96. The other two constructors in the middle are just passing their arguments through to the third one by its signature.

I don't think it needs to be in `options()`. I haven't set any global package settings that way yet, but if you think it would useful it's fine with me.

**Comment by mdsumner** [comment]
thanks!  I didn't understand that signature syntax before now ...

(I'm happy with not adding an option, I realized I will wrap the read with my own functions anyway with different defaults set for various cases)

**Comment by ctoney** [comment]
Sounds good. Currently, when we call the constructors from R they are recognized by the number of arguments. We can have up to six, but each one distinct by number of arguments. There is also the possibility to have alternate constructors with the same number of arguments recognized by different types, but requires additional steps (pointer to a function called a validator). All described in the vignette.

The thing about that vignette is that the section on exposed class is pretty short, but every time I go back and re-read it I notice something I didn't catch before.

I see you marked it ready for review. I hope to look at it later tonight but tomorrow at the latest. Thanks again.

**Review by ctoney — CHANGES_REQUESTED** [review]
Looks good. I added a few comments for minor edits in the documentation. Please do `devtools::document()` or equivalent one last time after making the changes so the .Rd files are rebuilt and committed. The `pkgdown` workflow will rebuild the online documentation. I'll update the changelog and dev version when it's merged.

Assuming you're agreeable, please add your information as contributor in `DESCRIPTION`. Besides this contribution, you've also been engaged with the package from its initial CRAN release. You've given feedback and suggestions that resulted in substantial improvements. The existing contributors/cph are "passive" contributors in the sense that they were added to Authors@R to comply with CRAN requirement. In that case, there is also attribution throughout in the documentation, code comments, and `inst/COPYRIGHTS` (per MIT license requirement, and the obvious obligation to anyway). I don't know if there is a protocol or convention to follow here, but it seems like "active" contributors might be listed first, i.e., contributors who do specific work with intent to add to this code or documentation. I'll defer to your judgement and preference on that, but is probably how I would do it.

Please note `CONTRIBUTING.md`, with respect to the public domain notice. I'm not asserting copyright on any of my work here. Please review `inst/COPYRIGHTS` for any concerns. Feel free now or in the future to assert copyright on any work you may contribute, by adding to that file, code comments, cph label, or whatever is appropriate. 

Just want to mention all that since the public domain aspect is atypical. Let me know if you have questions or concerns.

**Comment by mdsumner** [comment]
excellent thank you, I've done manually all those in one new commit

**Review by ctoney — APPROVED** [review]

**Comment by ctoney** [comment]
Nice addition. Thank you.

**Pull request merged**
